### PR TITLE
fix:hydrolib-core

### LINKED ISSUE #148: Revert incorrect hydrolib-core expected values
### Version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have checked that this bug exists on the latest version.

### Reproducible Example

-

### Current behaviour

In https://github.com/Deltares/hydromt_delft3dfm/pull/139 some mdu keywords were updated to new values.

Were empty before:
```
dzTop                      = -999                                                           # Z-layer thickness of layers above level Dztopuniabovez
floorLevTopLay             = -999                                                           # Floor level of top layer
dzTopUniAboveZ             = -999                                                           # Above this level layers will have uniform dzTop, below we use sigmaGrowthFactor
trtDef = .    # File (*.ttd) including trachytope definitions.
trtL   = .    # File (*.arl) including distribution of trachytope definitions.
outputDir                         = .      # Output directory of map-, his-, rst-, dat- and timingsfiles, default: DFM_OUTPUT_<modelname>. Set to . for no dir/current dir.
waqOutputDir                      = .      # Output directory of Water Quality files.
```

Not existent before:
```
fixedWeirTopFrictCoef = -999 # Uniform friction coefficient of the groyne part of fixed weirs [the unit depends on frictiontype].
```

Especially the `.` values are not expected, these might cause issues and they might not. This has been reported in hydrolib-core.

Todo:
- [ ] wait for https://github.com/Deltares/HYDROLIB-core/issues/703 fix (and new release)
- [ ] update minimal hydrolib-core version
- [ ] revert some defaults

### Desired behaviour

-

### Additional context

_No response_

### LINKED ISSUE #227: Add support for Python 3.14
Todo:

- [x] https://github.com/Deltares/HYDROLIB-core/issues/1010 >> probably requires https://github.com/Deltares/HYDROLIB-core/issues/732
- [x] await hydrolib-core release and update minimal version
- [x] add pixi 3.14 tests
- [x] cartopy does not have Python 3.14 wheels yet (latest release is from August 2025), they are working on this in https://github.com/SciTools/cartopy/issues/2612 >> https://github.com/SciTools/cartopy/pull/2641 >> https://github.com/SciTools/cartopy/pull/2653
- [x] remove fixed python version in pixi default env (or set to py314) and `pixi install` locally again (first delete lockfile and .pixi folder) (and run `pixi run test`)

### LINKED ISSUE #230: Update to hydrolib-core v1
Todo:
- [x] temporarily use hydrolib-core main branch for testing (1.0.0b5 with fixes)
- [x] fix any failing tests
- [x] https://github.com/Deltares/hydromt_delft3dfm/issues/234
- [x] https://github.com/Deltares/hydromt_delft3dfm/issues/138
- [x] check if we can safely assume that changing `forcing` to `forcing[0]` makes sense, what has changed? >> multiple forcingfiles are now supported, all forcings from all forcingfiles are parsed into a single dataframe now
- [x] none vs nan, if this is changed, also check for nan comparisons (like `nodeids = nodeids[nodeids != "nan"]` in the entire codebase)
- [x] [hydrolib-core 1.0 was released]
- [x] merge https://github.com/Deltares/hydromt_delft3dfm/pull/162 to main
- [x] merge main with this branch, fix conflicts
- [x] https://github.com/Deltares/hydromt_delft3dfm/issues/223
- [ ] https://github.com/Deltares/HYDROLIB-core/issues/853
- [ ] https://github.com/Deltares/HYDROLIB-core/issues/1053 >> https://github.com/Deltares/hydromt_delft3dfm/issues/148
- [ ] https://github.com/Deltares/HYDROLIB-core/issues/1051
- [ ] https://github.com/Deltares/HYDROLIB-core/issues/1055 >> revert the StructureModel specific comparison (or making it generic for all hydrolib-core objects)
- [ ] hydrolib-core v1: mdu misses previously empty keywords: dryPointsFile, landBoundaryFile, thinDamFile, fixedWeirFile, pillarFile, obsFile, crsFile
- [ ] release hydrolib-core
- [ ] require new minimal version >> `hydrolib-core>=1.0.1` (or so)
- [ ] consider picking up https://github.com/Deltares/hydromt_delft3dfm/issues/229 (many new nan-related-validation errors, so might be good to include immediately here)
- [ ] consider also initializing manhole_dn/manhole_up as None instead of ""
- [ ] update all reference files (many file headers are updated, but also nan to None)
- [ ] update changelog: `updated to hydrolib-core v1 (pydantic V2 API)`
- [ ] run example notebooks and maybe check resulting models

### PR DESCRIPTION
**Pull request by MAfarrag**
# Description

# Issues
- Fixes #

## Explanation
Explain how you addressed the bug/feature request, what choices you made and why.

## Checklist
- [ ] Updated tests or added new tests
- [ ] Branch is up to date with `main`
- [ ] Tests & pre-commit hooks pass
- [ ] Updated documentation if needed
- [ ] Updated changelog.rst if needed

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

**Comment by MAfarrag** [comment]
@veenstrajelmer can you check the assertion in the filing test, it is kind of weird assertion maybe you are familiar with it, 
so now all the hydrolib-core issues are passing

**Comment by veenstrajelmer** [comment]
@MAfarrag It is fixed, all tests are passing now. There was an error with comparing the structurefile, so I implemented a workaround to manually test for identicallness between the created and the reference model (they are identical). I expect that in hydromt v1, the equality checks are much better, so it can be done more efficient after merging https://github.com/Deltares/hydromt_delft3dfm/pull/162.

Thanks for all the hard work!

There are some things left todo in the issue, the first point about `forcing[0]` might also influence hydrolib-core: https://github.com/Deltares/hydromt_delft3dfm/issues/230

**Comment by veenstrajelmer** [comment]
@hboisgon could you please review the changes in this PR (specifically in `hydromt_delft3dfm/utils.py`)? We will wait with merging this PR until the hydromt v1 branch has been merged, but it is good to have this reviewed already.

File: hydromt_delft3dfm/utils.py [hunk 1]
```diff
@@ -690,7 +692,7 @@ def read_1dboundary(
     nodeids = nodeids[nodeids != "nan"]
     # Assume one forcing file (HydroMT writer) and read
     forcing = df.forcingfile.iloc[0]
-    df_forcing = pd.DataFrame([f.__dict__ for f in forcing.forcing])
+    df_forcing = pd.DataFrame([f.__dict__ for f in forcing[0].forcing])
```
Comment (veenstrajelmer) [inline hydromt_delft3dfm/utils.py]: I wonder why this would have changed. It worked in the past, so why should it not work anymore with hydrolib-core v1? It seems too fundamental to change this, but I might be mistaken

File: hydromt_delft3dfm/utils/io_utils.py [hunk 2]
```diff
@@ -148,6 +148,7 @@ def write_branches_gui(
     branches["branchtype"] = branches["branchtype"].replace(
         {"river": 0, "pipe": 2, "sewerconnection": 1}
     )
+    branches = branches.where(branches.notna(), None)
```
Comment (veenstrajelmer) [inline hydromt_delft3dfm/utils/io_utils.py]: nan to None

File: hydromt_delft3dfm/utils/io_utils.py [hunk 3]
```diff
@@ -395,11 +396,12 @@ def write_friction(gdf: gpd.GeoDataFrame, savedir: str) -> List[str]:
         ["frictionid", "frictionvalue", "frictiontype"]
     ]
     frictions = frictions.drop_duplicates().dropna(how="all")
+    frictions = frictions.where(frictions.notna(), None)
```
Comment (veenstrajelmer) [inline hydromt_delft3dfm/utils/io_utils.py]: nan to None

File: hydromt_delft3dfm/utils/io_utils.py [hunk 4]
```diff
@@ -395,11 +396,12 @@ def write_friction(gdf: gpd.GeoDataFrame, savedir: str) -> List[str]:
...
 
     friction_fns = []
     # create a new friction
     for i, row in frictions.iterrows():
-        if isinstance(row.frictionvalue, float) and not np.isnan(row.frictionvalue):
+        if row.frictionid is not None and row.frictionvalue is not None:
```
Comment (veenstrajelmer) [inline hydromt_delft3dfm/utils/io_utils.py:400]: nan to None

File: hydromt_delft3dfm/utils.py [hunk 5]
```diff
@@ -1054,7 +1056,7 @@ def read_2dboundary(df: pd.DataFrame, workdir: Path = Path.cwd()) -> xr.DataArra
 
     # Assume one forcing file (HydroMT writer) and read
     forcing = df.forcingfile
-    df_forcing = pd.DataFrame([f.__dict__ for f in forcing.forcing])
+    df_forcing = pd.DataFrame([f.__dict__ for f in forcing[0].forcing])
```
Comment (veenstrajelmer) [inline hydromt_delft3dfm/utils.py]: same

**Review by veenstrajelmer — COMMENTED** [review]
In the past in hydromt_delft3dfm, we skipped an important part of the validation. That still does not properly work, but I at least added the validation again and only ignored expected errors. This way, it is (sort of) avoided that the situation detoriates without us knowing. I merged it to main this morning and those tests passed, but I see you also merged main into your branch, and now this particular test fails. This is something to look into I think the main difference is a None/nan difference. I see that in this PR you convert nan to None, and I wonder if that is a good idea in this case. Let's discuss this next week.

File: hydromt_delft3dfm/utils.py (hunk 1 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils.py]: the forcing where returning differenet types `forcing | List[forcing]` and we had a lot of usage in hydrolib-core where we always had to check what is the type of returned value, is it single forcing or a list, and with type checker it was always flaged as an issue, so we decided to make the return value always as a list so you do not have to always check for what is returned value.

File: hydromt_delft3dfm/utils/io_utils.py (hunk 2 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils/io_utils.py]: change the new pandas `float("nan")` to the default value that hydrolib-core accepts, before pandas has this as None, so this is an issue of the pandas upgrade not hydrolib-core upgrade

File: hydromt_delft3dfm/utils/io_utils.py (hunk 3 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils/io_utils.py]: here the same

File: hydromt_delft3dfm/utils/io_utils.py (hunk 4 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils/io_utils.py:400]: same

File: hydromt_delft3dfm/utils.py (hunk 1 above)
Comment (veenstrajelmer) [reply hydromt_delft3dfm/utils.py]: Ok, fair point, but what if there is a forcing[1] also, that could happen right or am I wrong?

File: hydromt_delft3dfm/utils/io_utils.py (hunk 2 above)
Comment (veenstrajelmer) [reply hydromt_delft3dfm/utils/io_utils.py]: Ok, but this is a pandas 3 issue change right? This branch enforces `pandas<3` (to be resolved in https://github.com/Deltares/hydromt_delft3dfm/issues/229 to keep things clean).

File: hydromt_delft3dfm/utils.py (hunk 1 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils.py]: in most of the models, each boundary condition quantity in the ext file has only one bc file, however last year we enabled a feature that the boundary condition quantity could have multiple bc files, although we actually do not have any test case like that have multiple bc files, also read the comment added two lines above by `shartgring` on 19-5-2025.

Anyway now the comperhensive lise loops over the files, so if there are multiple files it will consider them all

File: hydromt_delft3dfm/utils/io_utils.py (hunk 2 above)
Comment (MAfarrag) [reply hydromt_delft3dfm/utils/io_utils.py]: well i thought pandas was already upgraded.

Comment (MAfarrag) [reply hydromt_delft3dfm/utils/io_utils.py]: well something changed instead of the filling the dataframe with None it is filled with float(""nan") and, the `Branch` class in `hydrolib-core` can take a None as a default value and that did not change

Comment (veenstrajelmer) [reply hydromt_delft3dfm/utils/io_utils.py]: I checked the resulting model files and all differences are making total sense. Thanks for your explanation.

**Review by veenstrajelmer — CHANGES_REQUESTED** [review]
Thanks for fixing everything. Everything now makes sense. There is one more thing to do, that is to replace the git link with `"hydrolib-core>1.0",` in pyproject.toml. So if you have time, please make the hydrolib-core release. After that, we would like to first merge the hydromt v1 branch since that is a much more extensive update and we would like to avoid merge conflicts with that. Just so you know. The full list of todo's for this PR is in the issue: https://github.com/Deltares/hydromt_delft3dfm/issues/230. I will pick this up asap.

**Review by veenstrajelmer — APPROVED** [review]
